File: docs/tests/test_doc.py
import sys
CYTNX_ROOT = '/home/hunghaoti/Libraries/Cytnx_lib'
sys.path.insert(0, CYTNX_ROOT)
import cytnx
import pytest
import time

ex_code_path = '../code/python/'
test_codes_path = ex_code_path + 'doc_codes/'
output_path = ex_code_path + 'outputs/'

def needGPUTest(obj):
    #if need to test cuda, set us True
    obj.__test__ = False
    return obj

def get_redirect_code(test_name):
    str_code = '''
captured = capsys.readouterr()
if captured.out:
    text_file = open(output_path + \'''' + test_name + '''.out', 'w')
    text_file.write(captured.out)
    text_file.close()
'''
    return str_code

def excute_all(test_names, capsys):
    codes = ''
    for test_name in test_names:
        code = open(test_codes_path + test_name + '.py').read() + \
               get_redirect_code(test_name)
        codes = codes + code
    exec(codes)


def excute_and_output(test_name, capsys):
    exec(open(test_codes_path + test_name + '.py').read())
    captured = capsys.readouterr()
    text_file = open(output_path + test_name + '.out', 'w')
    text_file.write(captured.out)
    text_file.close()

# test case main

# Objects behavior
def test_guide_behavior_assign(capsys):
    excute_and_output('guide_behavior_assign', capsys)

def test_guide_behavior_clone(capsys):
    excute_and_output('guide_behavior_clone', capsys)

# Device
def test_guide_Device_Ncpus(capsys):
    excute_and_output('guide_Device_Ncpus', capsys)

def test_guide_Device_Ngpus(capsys):
    excute_and_output('guide_Device_Ngpus', capsys)

def test_guide_Device_property(capsys):
    excute_and_output('guide_Device_property', capsys)

# Tensor
#  -Creating a Tesnor
def test_guide_basic_obj_Tensor_1_create_zeros(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_zeros', capsys)

def test_guide_basic_obj_Tensor_1_create_diff_ways(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_diff_ways', capsys)

def test_guide_basic_obj_Tensor_1_create_rand(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_rand', capsys)

@needGPUTest
def test_guide_basic_obj_Tensor_1_create_zeros_cuda(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_zeros_cuda', capsys)

def test_guide_basic_obj_Tensor_1_create_astype(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_astype', capsys)

@needGPUTest
def test_guide_basic_obj_Tensor_1_create_to(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_to', capsys)

def test_guide_basic_obj_Tensor_1_create_from_storage(capsys):
    excute_and_output('guide_basic_obj_Tensor_1_create_from_storage', capsys)

# -Manipulating Tensors
def test_guide_basic_obj_Tensor_2_manip_reshape(capsys):
    excute_and_output('guide_basic_obj_Tensor_2_manip_reshape', capsys)

def test_guide_basic_obj_Tensor_2_manip_reshape_(capsys):
    excute_and_output('guide_basic_obj_Tensor_2_manip_reshape_', capsys)

def test_guide_basic_obj_Tensor_2_manip_permute(capsys):
    excute_and_output('guide_basic_obj_Tensor_2_manip_permute', capsys)

def test_guide_basic_obj_Tensor_2_manip_contiguous(capsys):
    excute_and_output('guide_basic_obj_Tensor_2_manip_contiguous', capsys)

# -Accessing elements
def test_guide_basic_obj_Tensor_3_access_slice_get(capsys):
    excute_and_output('guide_basic_obj_Tensor_3_access_slice_get', capsys)

def test_guide_basic_obj_Tensor_3_access_item(capsys):
    excute_and_output('guide_basic_obj_Tensor_3_access_item', capsys)

def test_guide_basic_obj_Tensor_3_access_slice_set(capsys):
    excute_and_output('guide_basic_obj_Tensor_3_access_slice_set', capsys)

# -Tensor arithmetic
def test_guide_basic_obj_Tensor_4_arithmetic_tensor_scalar(capsys):
    excute_and_output('guide_basic_obj_Tensor_4_arithmetic_tensor_scalar', capsys)

def test_guide_basic_obj_Tensor_4_arithmetic_tensor_tensor(capsys):
    excute_and_output('guide_basic_obj_Tensor_4_arithmetic_tensor_tensor', capsys)

# -To/From numpy.array
def test_guide_basic_obj_Tensor_5_numpy_cytnx2numpy(capsys):
    excute_and_output('guide_basic_obj_Tensor_5_numpy_cytnx2numpy', capsys)

def test_guide_basic_obj_Tensor_5_numpy_numpy2cytnx(capsys):
    excute_and_output('guide_basic_obj_Tensor_5_numpy_numpy2cytnx', capsys)

# -Appending elements
def test_guide_basic_obj_Tensor_6_app_scalar(capsys):
    excute_and_output('guide_basic_obj_Tensor_6_app_scalar', capsys)

def test_guide_basic_obj_Tensor_6_app_tensor(capsys):
    excute_and_output('guide_basic_obj_Tensor_6_app_tensor', capsys)

# -Save/Load a Tensor
def test_guide_basic_obj_Tensor_7_io_Save(capsys):
    excute_and_output('guide_basic_obj_Tensor_7_io_Save', capsys)

def test_guide_basic_obj_Tensor_7_io_Load(capsys):
    excute_and_output('guide_basic_obj_Tensor_7_io_Load', capsys)

# -When will data be coped
def test_guide_basic_obj_Tensor_8_cp_assign(capsys):
    excute_and_output('guide_basic_obj_Tensor_8_cp_assign', capsys)

def test_guide_basic_obj_Tensor_8_cp_clone(capsys):
    excute_and_output('guide_basic_obj_Tensor_8_cp_clone', capsys)

def test_guide_basic_obj_Tensor_8_cp_permute(capsys):
    test_names = ['guide_basic_obj_Tensor_8_cp_permute-1', \
                  'guide_basic_obj_Tensor_8_cp_permute-2', \
                  'guide_basic_obj_Tensor_8_cp_permute-3']
    excute_all(test_names, capsys)

def test_guide_basic_obj_Tensor_8_cp_contiguous(capsys):
    test_names = ['guide_basic_obj_Tensor_8_cp_contiguous-1', \
                  'guide_basic_obj_Tensor_8_cp_contiguous-2']
    excute_all(test_names, capsys)

# Storage
# -Creating a Storage
def test_guide_basic_obj_Storage_1_create_create(capsys):
    excute_and_output('guide_basic_obj_Storage_1_create_create', capsys)

def test_guide_basic_obj_Storage_1_create_astype(capsys):
    excute_and_output('guide_basic_obj_Storage_1_create_astype', capsys)

@needGPUTest
def test_guide_basic_obj_Storage_1_create_to(capsys):
    excute_and_output('guide_basic_obj_Storage_1_create_to', capsys)

def test_guide_basic_obj_Storage_1_create_get_storage(capsys):
    excute_and_output('guide_basic_obj_Storage_1_create_get_storage', capsys)

def test_guide_basic_obj_Storage_1_create_contiguous_check(capsys):
    excute_and_output('guide_basic_obj_Storage_1_create_contiguous_check', capsys)

# -Accessing elements
def test_guide_basic_obj_Storage_2_access_access(capsys):
    excute_and_output('guide_basic_obj_Storage_2_access_access', capsys)

# -Increasing size
def test_guide_basic_obj_Storage_3_expand_append(capsys):
    excute_and_output('guide_basic_obj_Storage_3_expand_append', capsys)

def test_guide_basic_obj_Storage_3_expand_resize(capsys):
    excute_and_output('guide_basic_obj_Storage_3_expand_resize', capsys)

# -Save/Load a storage
def test_guide_basic_obj_Storage_5_io_Save(capsys):
    excute_and_output('guide_basic_obj_Storage_5_io_Save', capsys)

def test_guide_basic_obj_Storage_5_io_Load(capsys):
    excute_and_output('guide_basic_obj_Storage_5_io_Load', capsys)

def test_guide_basic_obj_Storage_5_io_from_to_file(capsys):
    excute_and_output('guide_basic_obj_Storage_5_io_from_to_file', capsys)

# UniTensor
# -Print and display
def test_guide_uniten_print(capsys):
    test_names = ['guide_uniten_print_init', \
                  'guide_uniten_print_print_diagram', \
                  'guide_uniten_print_set_name', \
                  'guide_uniten_print_print_block', \
                  'guide_uniten_print_sym_print', \
                  'guide_uniten_print_sym_print_block']
    excute_all(test_names, capsys)

# -Creating a UniTensor
def test_guide_uniten_create_from_tensor(capsys):
    excute_and_output('guide_uniten_create_from_tensor', capsys)

def test_guide_uniten_create_complex(capsys):
    test_names = ['guide_uniten_create_complex', \
                  'guide_uniten_create_print_diagram']
    excute_all(test_names, capsys)

def test_guide_uniten_create_scratch(capsys):
    excute_and_output('guide_uniten_create_scratch', capsys)

# -Changing labels
def test_guide_uniten_labels_relabel(capsys):
    test_names = ['guide_uniten_labels_relabel_', \
                  'guide_uniten_labels_relabel']
    excute_all(test_names, capsys)

# -Bond
def test_guide_uniten_bond_create(capsys):
    excute_and_output('guide_uniten_bond_create', capsys)

def test_guide_uniten_bond_redirect(capsys):
    excute_and_output('guide_uniten_bond_redirect', capsys)

def test_guide_uniten_bond_symobj(capsys):
    excute_and_output('guide_uniten_bond_symobj', capsys)

def test_guide_uniten_bond_sym_bond(capsys):
    test_names = ['guide_uniten_bond_sym_bond', \
                 'guide_uniten_bond_multi_sym_bond', \
                 'guide_uniten_bond_combine', \
                 'guide_uniten_bond_combine_no_grp']
    excute_all(test_names, capsys)

# -Tagged UniTensor
def test_guide_uniten_tagged(capsys):
    with pytest.raises(Exception) as e_info:
        test_names = ['guide_uniten_tagged_init', \
                      'guide_uniten_tagged_contract']
        excute_all(test_names, capsys)
    print(str(e_info.value))
    test_name = 'guide_uniten_tagged_contract'
    exec(get_redirect_code(test_name))

# -UniTensor with Symmetries
def test_guide_uniten_symmetric(capsys):
    test_names = ['guide_uniten_symmetric_create', \
                  'guide_uniten_symmetric_print_blocks']
    excute_all(test_names, capsys)

# -Accessing the blocks(s)
def test_guide_uniten_blocks_get_block(capsys):
    excute_and_output('guide_uniten_blocks_get_block', capsys)

def test_guide_uniten_blocks_put_get_block(capsys):
    test_names = ['guide_uniten_blocks_init', \
                  'guide_uniten_blocks_get_block_qidx', \
                  'guide_uniten_blocks_get_block_bkidx', \
                  'guide_uniten_blocks_get_blocks_', \
                  'guide_uniten_blocks_put_block_qidx', \
                  'guide_uniten_blocks_put_block_bkidx']
    excute_all(test_names, capsys)

# -Get/set UniTensor element
def test_guide_uniten_elements_at_get(capsys):
    excute_and_output('guide_uniten_elements_at_get', capsys)

def test_guide_uniten_elements_at_set(capsys):
    excute_and_output('guide_uniten_elements_at_set', capsys)

def test_guide_uniten_elements_sym(capsys):
    with pytest.raises(Exception) as e_info:
        test_names = ['guide_uniten_elements_init_sym', \
                      'guide_uniten_elements_at_qidx', \
                      'guide_uniten_elements_at_non_exist']
        excute_all(test_names, capsys)
    print(str(e_info.value))
    test_name = 'guide_uniten_elements_at_non_exist'
    exec(get_redirect_code(test_name))

def test_guide_uniten_elements_exists(capsys):
    test_names = ['guide_uniten_elements_init_sym', \
                  'guide_uniten_elements_exists']
    excute_all(test_names, capsys)

# -Manipulate UniTensor
def test_guide_uniten_manipulation_permute(capsys):
    excute_and_output('guide_uniten_manipulation_permute', capsys)

def test_guide_uniten_manipulation_reshape(capsys):
    excute_and_output('guide_uniten_manipulation_reshape', capsys)

def test_guide_uniten_manipulation_combine(capsys):
    excute_and_output('guide_uniten_manipulation_combine', capsys)

# Save/Load a UniTensor
def test_guide_uniten_io_Save(capsys):
    excute_and_output('guide_uniten_io_Save', capsys)
    time.sleep(0.01) # wait to save field, then next test can load

def test_guide_uniten_io_Load(capsys):
    excute_and_output('guide_uniten_io_Load', capsys)

# Contraction
# -Network
def test_guide_contraction_network_launch(capsys):
    test_names = ['guide_contraction_network_PutUniTensor', \
                  'guide_contraction_network_launch']
    excute_all(test_names, capsys)

def test_guide_contraction_network_FromString(capsys):
    excute_and_output('guide_contraction_network_FromString', capsys)

def test_guide_contraction_network_label_ord(capsys):
    test_names = ['guide_contraction_network_label_ord-1', \
                  'guide_contraction_network_label_ord-2', \
                  'guide_contraction_network_label_ord-3']
    excute_all(test_names, capsys)

# -Contract
def test_guide_contraction_contract_Contract(capsys):
    excute_and_output('guide_contraction_contract_Contract', capsys)

def test_guide_contraction_contract_relabes(capsys):
    excute_and_output('guide_contraction_contract_relabels', capsys)

def test_guide_contraction_contract_Contracts(capsys):
    excute_and_output('guide_contraction_contract_Contracts', capsys)

# -ncon
def test_guide_contraction_ncon_ncon(capsys):
    excute_and_output('guide_contraction_ncon_ncon', capsys)

# Iterative solver
# -LinOp class
def test_guide_itersol_LinOp_Dot(capsys):
    excute_and_output('guide_itersol_LinOp_Dot', capsys)

def test_guide_itersol_LinOp_inherit(capsys):
    excute_and_output('guide_itersol_LinOp_inherit', capsys)

def test_guide_itersol_LinOp_matvec(capsys):
    test_names = ['guide_itersol_LinOp_matvec', \
                  'guide_itersol_LinOp_demo']
    excute_all(test_names, capsys)

def test_guide_itersol_LinOp_sparse_mv(capsys):
    excute_and_output('guide_itersol_LinOp_sparse_mv', capsys)

def test_guide_itersol_LinOp_sparse_mv_elem(capsys):
    excute_and_output('guide_itersol_LinOp_sparse_mv_elem', capsys)

# -Lanczos solver
def test_guide_itersol_Lanczos_Lanczos(capsys):
    excute_and_output('guide_itersol_Lanczos_Lanczos', capsys)

# linalg extension
# -Tensor decomposistion
def test_guide_xlinalg_Svd(capsys):
    test_names = ['guide_xlinalg_Svd', \
                  'guide_xlinalg_Svd_verify']
    excute_all(test_names, capsys)

File: docs/tests/test_doc.cpp
#include "cytnx.hpp"
#include <gtest/gtest.h>

// redirect standard output: if you need to output to another file, you can use these
//     macro to redirect the standart output
#define OUT_REDIRECT_BASE \
  coutbuf = cout.rdbuf(); \
  cout.rdbuf(output_file_o.rdbuf());  // redirect std::cout to file

#define OUT_REDIRECT                                                                 \
  string suite_name = testing::UnitTest::GetInstance()->current_test_info()->name(); \
  \ 
	output_file_o.open(output_dir + suite_name + ".out");                        \
  OUT_REDIRECT_BASE

#define OUT_REDIRECT_FILE(OUT_FILE)                   \
  output_file_o.open(output_dir + OUT_FILE + ".out"); \
  OUT_REDIRECT_BASE

#define OUT_RESET      \
  cout.rdbuf(coutbuf); \
  output_file_o.close();

using namespace cytnx;
using namespace std;
using namespace testing;

namespace DocTest {
  const string output_dir = "../../code/cplusplus/outputs/";

  // output redirect
  streambuf *coutbuf = nullptr;
  ofstream output_file_o;

  // 1. Objects behavior
  // 1.1. Everyting is reference
  TEST(Doc, guide_behavior_assign) {
    OUT_REDIRECT
    #include "guide_behavior_assign.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_behavior_clone) {
    OUT_REDIRECT
    #include "guide_behavior_clone.cpp"
    OUT_RESET
  }

  // 1.2 clone
  // 2. Device
  // 2.1 Number of threads
  TEST(Doc, guide_Device_Ncpus) {
    OUT_REDIRECT
    #include "guide_Device_Ncpus.cpp"
    OUT_RESET
  }

  // 2.2 Number of GPUs
  TEST(Doc, guide_Device_Ngpus) {
    OUT_REDIRECT
    #include "guide_Device_Ngpus.cpp"
    OUT_RESET
  }

  // 3. Tensor
  // 3.1 Creating a Tensor
  // 3.1.1 Initialized Tensor
  TEST(Doc, guide_basic_obj_Tensor_1_create_zeros) {
    #include "guide_basic_obj_Tensor_1_create_zeros.cpp"
  }

  TEST(Doc, guide_basic_obj_Tensor_1_create_diff_ways) {
    #include "guide_basic_obj_Tensor_1_create_diff_ways.cpp"
  }

  // 3.1.2 Random Tensor
  TEST(Doc, guide_basic_obj_Tensor_1_create_rand) {
    #include "guide_basic_obj_Tensor_1_create_rand.cpp"
  }

// 3.1.3 Tensor with different dtype and device
#ifdef UNI_GPU
  TEST(Doc, guide_basic_obj_Tensor_1_create_zeros_cuda) {
    #include "guide_basic_obj_Tensor_1_create_zeros_cuda.cpp"
  }
#endif

  // 3.1.4 Type conversion
  TEST(Doc, guide_basic_obj_Tensor_1_create_astype) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_1_create_astype.cpp"
    OUT_RESET
  }

// 3.1.5 Transfer between devices
#ifdef UNI_GPU
  TEST(Doc, guide_basic_obj_Tensor_1_create_to) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_1_create_to.cpp"
    OUT_RESET
  }
#endif

  // 3.1.6 Tensor from Storage
  TEST(Doc, guide_basic_obj_Tensor_1_create_from_storage) {
    #include "guide_basic_obj_Tensor_1_create_from_storage.cpp"
  }

  // 3.2 Manipulating Tensors
  // 3.2.1 reshape
  TEST(Doc, guide_basic_obj_Tensor_2_manip_reshape) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_2_manip_reshape.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Tensor_2_manip_reshape_) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_2_manip_reshape_.cpp"
    OUT_RESET
  }

  // 3.2.2 permute
  TEST(Doc, guide_basic_obj_Tensor_2_manip_permute) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_2_manip_permute.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Tensor_2_manip_contiguous) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_2_manip_contiguous.cpp"
    OUT_RESET
  }

  // 3.3 Accessing elements
  // 3.3.1 Get elements
  TEST(Doc, guide_basic_obj_Tensor_3_access_slice_get) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_3_access_slice_get.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Tensor_3_access_item) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_3_access_item.cpp"
    OUT_RESET
  }

  // 3.3.2 Get elements
  TEST(Doc, guide_basic_obj_Tensor_3_access_slice_set) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_3_access_slice_set.cpp"
    OUT_RESET
  }

  // 3.3.3 Low-level API (C++ only)
  /*
  TEST(Doc, guide_basic_obj_Tensor_3_access_c_accessor) {
    #include "guide_basic_obj_Tensor_3_access_c_accessor.cpp"
  }

  TEST(Doc, guide_basic_obj_Tensor_3_access_c_operator) {
    #include "guide_basic_obj_Tensor_3_access_c_operator.cpp"
  }

  TEST(Doc, guide_basic_obj_Tensor_3_access_c_get_set) {
  }
    #include "guide_basic_obj_Tensor_3_access_c_get_set.cpp"
  */

  // 3.4 Tensor arithmetic
  // 3.4.2 Tensor-scalar arithmetic
  TEST(Doc, guide_basic_obj_Tensor_4_arithmetic_tensor_scalar) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_4_arithmetic_tensor_scalar.cpp"
    OUT_RESET
  }

  // 3.4.3 Tensor-Tensor arithmetic
  TEST(Doc, guide_basic_obj_Tensor_4_arithmetic_tensor_tensor) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_4_arithmetic_tensor_tensor.cpp"
    OUT_RESET
  }

  // 3.4.4 Equivalent APIs (C++ only)
  TEST(Doc, guide_basic_obj_Tensor_4_arithmetic_Add) {
    #include "guide_basic_obj_Tensor_4_arithmetic_Add.cpp"
  }

  // 3.6 Appending elements
  TEST(Doc, guide_basic_obj_Tensor_6_app_scalar) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_6_app_scalar.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Tensor_6_app_tensor) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_6_app_tensor.cpp"
    OUT_RESET
  }

  // 3.7 Save/Load
  // 3.7.1 Save a Tensor
  TEST(Doc, guide_basic_obj_Tensor_7_io_Save) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_7_io_Save.cpp"
    OUT_RESET
  }

  // 3.7.2 Load a Tensor
  TEST(Doc, guide_basic_obj_Tensor_7_io_Load) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_7_io_Load.cpp"
    OUT_RESET
  }

  // 3.8 When will data be copied?
  // 3.8.1 Reference to & Copy of objects
  TEST(Doc, guide_basic_obj_Tensor_8_cp_assign) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_8_cp_assign.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Tensor_8_cp_clone) {
    OUT_REDIRECT
    #include "guide_basic_obj_Tensor_8_cp_clone.cpp"
    OUT_RESET
  }

  // 3.8.2 Permute
  TEST(Doc, guide_basic_obj_Tensor_8_cp_permute) {
    OUT_REDIRECT_FILE("guide_basic_obj_Tensor_8_cp_permute-1")
    #include "guide_basic_obj_Tensor_8_cp_permute-1.cpp"
    OUT_RESET OUT_REDIRECT_FILE("guide_basic_obj_Tensor_8_cp_permute-2")
    #include "guide_basic_obj_Tensor_8_cp_permute-2.cpp"
    OUT_RESET OUT_REDIRECT_FILE("guide_basic_obj_Tensor_8_cp_permute-3")
    #include "guide_basic_obj_Tensor_8_cp_permute-3.cpp"
    OUT_RESET
  }

  // 3.8.3 Contiguous
  TEST(Doc, guide_basic_obj_Tensor_8_cp_contiguous){
    OUT_REDIRECT_FILE("guide_basic_obj_Tensor_8_cp_contiguous-1")
    #include "guide_basic_obj_Tensor_8_cp_contiguous-1.cpp"
    OUT_RESET OUT_REDIRECT_FILE("guide_basic_obj_Tensor_8_cp_contiguous-2")
    #include "guide_basic_obj_Tensor_8_cp_contiguous-2.cpp"
    OUT_RESET
  }

  // 4. Storage
  // 4.1 Creating a Storage
  TEST(Doc, guide_basic_obj_Storage_1_create_create) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_1_create_create.cpp"
    OUT_RESET
  }

  // 4.1.1 Type conversion
  TEST(Doc, guide_basic_obj_Storage_1_create_astype) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_1_create_astype.cpp"
    OUT_RESET
  }

// 4.1.2 Transfer between devices
#ifdef UNI_GPU
  TEST(Doc, guide_basic_obj_Storage_1_create_to) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_1_create_to.cpp"
    OUT_RESET
  }
#endif

  // 4.1.3 Get Storage of Tensor
  TEST(Doc, guide_basic_obj_Storage_1_create_get_storage) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_1_create_get_storage.cpp"
    OUT_RESET
  }

  // 4.2 Accessing elements
  // 4.2.1 Get/Set elements
  TEST(Doc, guide_basic_obj_Storage_2_access_access) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_2_access_access.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Storage_2_access_at) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_2_access_at.cpp"
    OUT_RESET
  }

  // 4.2.2 Get raw-pointer (C++ only)
  TEST(Doc, guide_basic_obj_Storage_2_access_ptr_T) {
    #include "guide_basic_obj_Storage_2_access_ptr_T.cpp"
  }

  TEST(Doc, guide_basic_obj_Storage_2_access_ptr_void) {
    #include "guide_basic_obj_Storage_2_access_ptr_void.cpp"
  }

  // 4.3 Increase size
  // 4.3.1 append
  TEST(Doc, guide_basic_obj_Storage_3_expand_append) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_3_expand_append.cpp"
    OUT_RESET
  }

  // 4.3.2 resize
  TEST(Doc, guide_basic_obj_Storage_3_expand_resize) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_3_expand_resize.cpp"
    OUT_RESET
  }

// 4.4 From/To C++ .vector
#ifdef UNI_GPU
  TEST(Doc, guide_basic_obj_Storage_4_vec_from_vec) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_4_vec_from_vec.cpp"
    OUT_RESET
  }
#endif

  TEST(Doc, guide_basic_obj_Storage_4_vec_to_vec) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_4_vec_to_vec.cpp"
    OUT_RESET
  }

  // 4.5 Save/Load
  // 4.5.1 Save a Storage
  TEST(Doc, guide_basic_obj_Storage_5_io_Save) {
    #include "guide_basic_obj_Storage_5_io_Save.cpp"
  }

  // 4.5.2 Load a Storage
  TEST(Doc, guide_basic_obj_Storage_5_io_Load) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_5_io_Load.cpp"
    OUT_RESET
  }

  // 4.5.3 Save & load from/to binary
  TEST(Doc, guide_basic_obj_Storage_5_io_from_to_file) {
    OUT_REDIRECT
    #include "guide_basic_obj_Storage_5_io_from_to_file.cpp"
    OUT_RESET
  }

  // 5. Scalar
  // 5.1 Define/Declare a Scalar
  TEST(Doc, guide_basic_obj_Scalar_from_c_var) {
    OUT_REDIRECT
    #include "guide_basic_obj_Scalar_from_c_var.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Scalar_create) {
    OUT_REDIRECT
    #include "guide_basic_obj_Scalar_create.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_basic_obj_Scalar_to_c_var) {
    OUT_REDIRECT
    #include "guide_basic_obj_Scalar_to_c_var.cpp"
    OUT_RESET
  }

  // 5.2 Change date type
  TEST(Doc, guide_basic_obj_Scalar_astype) {
    OUT_REDIRECT
    #include "guide_basic_obj_Scalar_astype.cpp"
    OUT_RESET
  }

  // 5.3 Application scenarios
  TEST(Doc, guide_basic_obj_Scalar_application) {
    OUT_REDIRECT
    #include "guide_basic_obj_Scalar_application.cpp"
    OUT_RESET
  }

  // 7. UniTensor
  // 7.4 Bond
  // 7.4.1 Symmetry object
  TEST(Doc, guide_uniten_bond_symobj) {
    OUT_REDIRECT
    #include "guide_uniten_bond_symobj.cpp"
    OUT_RESET
  }

  // 7.4.2 Creating Bonds with quantum numbers
  TEST(Doc, guide_uniten_bond_sym_bond) {
    OUT_REDIRECT
    #include "guide_uniten_bond_sym_bond.cpp"
    OUT_RESET
  }

  // 7.8 Get/set UniTensor element
  // 7.8.1 UniTensor without symmetries
  TEST(Doc, guide_uniten_elements_at_get) {
    OUT_REDIRECT
    #include "guide_uniten_elements_at_get.cpp"
    OUT_RESET
  }

  TEST(Doc, guide_uniten_elements_at_set) {
    OUT_REDIRECT
    #include "guide_uniten_elements_at_set.cpp"
    OUT_RESET
  }

  // 7.8.1 UniTensor with symmetries
  TEST(Doc, user_guide_7_8_2_ex1) {
    // need to add
  }

  // 8. Contraction
  // 8.1 Network
  // 8.1.2 Put UniTensors and Launch
  TEST(Doc, guide_contraction_network_launch) {
    OUT_REDIRECT_FILE("guide_contraction_network_PutUniTensor")
    #include "guide_contraction_network_PutUniTensor.cpp"
    OUT_RESET
    #include "guide_contraction_network_launch.cpp"
  }

  // 10. Iterative solver
  // 10.1 LinOp class
  TEST(Doc, guide_itersol_LinOp_Dot) {
    OUT_REDIRECT
    #include "guide_itersol_LinOp_Dot.cpp"
    OUT_RESET
  }

  // 10.1.2 Inherit the LinOp class
  TEST(Doc, guide_itersol_LinOp_matvec) {
    OUT_REDIRECT
    #include "guide_itersol_LinOp_matvec.cpp"
    OUT_RESET
  }

  // 10.2 Lanczos solver
  TEST(Doc, guide_itersol_Lanczos_Lanczos) {
    OUT_REDIRECT
    #include "guide_itersol_Lanczos_Lanczos.cpp"
    OUT_RESET
  }

}  // namespace DocTest
